### PR TITLE
Merge Support: Populate start time for changeFeed

### PR DESCRIPTION
**Pull request by xinlian12**
Part of the effort of adding Merge support in Java SDK. Main working item: https://github.com/Azure/azure-sdk-for-java/issues/32131

This change is related to working item: https://github.com/Azure/azure-sdk-for-java/issues/32164

**Issue:**
When a merge happens, the child partition will contain documents ordered by LSN but the _ts / creation time of those documents might not be sequential. So when reading the Change Feed by LSN, you can encounter documents with a lower _ts. This becomes problematic when a user consumes the Change Feed using Start Time because:

The first request sends the desired Start Time.
Backend calculates an LSN and returns to the client
Client uses that LSN moving forward.
Because of the previous merge, potentially can receive documents that happened after the Start Time because they have higher LSN than returned in step 2.

This only impacts Change Feed Processor and Change Feed Query/Pull model that are started after a merge happened in a partition, not running ones, because when a Partition is merged we currently apply EPK filtering, and with that filtering, we would not get documents from the other parent Partition that could be out of time.

**Changes:**
Modifying the SDK code that performs Change Feed requests to send the new header obtained from the current continuation/lease

Starting point | Header sent on first request | Headers being sent on following requests(Current) | Headers being sent on following requests(In this PR)
-- | -- | -- | --
startFromBeginning | N/A | "If-None-Match" -> "LSN" | "If-None-Match" -> "LSN"
startFromNow | "If-None-Match" -> "*" | "If-None-Match" -> "LSN" | "If-None-Match" -> "LSN"
startFromPointInTime | "If-Modified-Since" -> startTime | "If-None-Match" -> "LSN" | "If-None-Match" -> "LSN""If-Modified-Since" -> startTime

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/common/ChangeFeedStateV1.java [hunk 1]
```diff
@@ -114,6 +114,23 @@ private void populateEffectiveRangeAndStartFromSettingsToRequest(RxDocumentServi
...
+
+        checkNotNull(initialStartFrom, "Argument 'initialStartFrom' should not be null");
+        checkNotNull(effectiveStartFrom, "Argument 'effectiveStartFromSettings' should not be null");
+        checkNotNull(request, "Argument 'request' should not be null");
+
+        // When a merge happens, the child partition will contain documents ordered by LSN but the _ts/creation time
+        // of the documents may not be sequential. So when reading the changeFeed by LSN, it is possible to encounter documents with lower _ts.
+        // In order to guarantee we always get the documents after customer's point start time, we will need to always pass the start time in the header.
+        // NOTE: the sequence of calling populate request order matters here, as both can try to populate the same header, effective ones will win
+        initialStartFrom.populateRequest(request, this.mode);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/common/ChangeFeedStateV1.java:133]: I am still waiting to confirm with backend about the capabilities header contract

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/ChangeFeedStateTest.java [hunk 2]
```diff
@@ -5,22 +5,47 @@
...
+            { ChangeFeedMode.INCREMENTAL, ChangeFeedStartFromTypes.POINT_IN_TIME, true },
+            { ChangeFeedMode.INCREMENTAL, ChangeFeedStartFromTypes.BEGINNING, false },
+            { ChangeFeedMode.INCREMENTAL, ChangeFeedStartFromTypes.NOW, false },
+            { ChangeFeedMode.INCREMENTAL, ChangeFeedStartFromTypes.POINT_IN_TIME, false },
+            { ChangeFeedMode.FULL_FIDELITY, ChangeFeedStartFromTypes.BEGINNING, true },
+            { ChangeFeedMode.FULL_FIDELITY, ChangeFeedStartFromTypes.NOW, true },
+            { ChangeFeedMode.FULL_FIDELITY, ChangeFeedStartFromTypes.POINT_IN_TIME, true },
+            { ChangeFeedMode.FULL_FIDELITY, ChangeFeedStartFromTypes.BEGINNING, false },
+            { ChangeFeedMode.FULL_FIDELITY, ChangeFeedStartFromTypes.NOW, false },
+            { ChangeFeedMode.FULL_FIDELITY, ChangeFeedStartFromTypes.POINT_IN_TIME, false },
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/ChangeFeedStateTest.java:45]: Full fidelity doesn't support point in time.

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/ChangeFeedStateTest.java:45]: Are we testing only to create ChangeFeedState?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/epkversion/ServiceItemLeaseV1.java [hunk 3]
```diff
@@ -150,17 +149,11 @@ public ChangeFeedState getContinuationState(String containerRid, ChangeFeedMode
...
-        String token = changeFeedState.getContinuation().getCurrentContinuationToken().getToken();
-        //  This token has extra quotes
-        token = token.replace("\"", "");
-
         return new ChangeFeedStateV1(
             containerRid,
             this.feedRangeInternal,
             changeFeedMode,
-            ChangeFeedStartFromInternal.createFromETagAndFeedRange(token, this.feedRangeInternal),
+            changeFeedState.getStartFromSettings(),
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/epkversion/ServiceItemLeaseV1.java:156]: why do we need to change this?

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/ChangeFeedStateTest.java (hunk 2 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/ChangeFeedStateTest.java:45]: With this change, we should also test through our live tests for FullFidelityChangeFeedTest .

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @xinlian12

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/epkversion/ServiceItemLeaseV1.java (hunk 3 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/epkversion/ServiceItemLeaseV1.java:156]: I think the startFromSettings inside the ChangeFeedState should always reflect the original start from and we are depending on it to always send the start time.

Currently, if we resume from the existing lease or split happens, then we lost the original startFrom info, which I think is a bug

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/epkversion/ServiceItemLeaseV1.java:156]: i think the startFrom should always track the original startFrom settings of the changeFeedState and it is critical for us to keep sending the startFrom header properly for the following request. 

Currently, without this change, the original startFrom information will be missed when we resume processing from current lease or when a split happens

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/ChangeFeedStateTest.java (hunk 2 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/ChangeFeedStateTest.java]: Thanks, removed the unapplicable settings. 

The testing for this one is a little bit tricky, as we need to examine the final headers we are going to sent to backend. I did local e2e test for this change to check the point in time info has always been sent through the header. Will need to think about about how to do header validation in our tests.

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/ChangeFeedStateTest.java]: And FullFidelityChangeFeedTest, it should not impact it much, as point in time is not applicable.

**Pull request merged**